lib/http/response-validator/validationStreams.js

```diff
@@ -0,0 +1,131 @@
+var CHECKSUM_BYTE_LEN = 16;
+
+module.exports = {
+  IntegrityCheckerStream: IntegrityCheckerStream,
+  ContentLengthCheckerStream: ContentLengthCheckerStream,
+}
+
+require('util').inherits(IntegrityCheckerStream, TransformStream);
+require('util').inherits(ContentLengthCheckerStream, TransformStream);
+
+/**
+ * Only runnable in Node
+ * A transform stream to perform trailing md5 hash validation.
+ * Whether to perform this validation is based on 'x-amz-tranfer-encoding'
+ * response header.
+ * @api private
+ */
+function IntegrityCheckerStream(transferEncoding, contentLength) {
+  if (!(this instanceof TransformStream)) {
+    return new TransformStream();
+  }
+  TransformStream.call(this);
+  if (transferEncoding !== 'append-md5') {
+    process.nextTick(function(stream, encoding) {
+      stream.emit(
+        'error',
+        AWS.util.error(
+          new Error('Cannot validate response header: ' + encoding + ', expect \'append-md5\''),
+          {code: 'UnsupportedHashingAlgorithm'}
+        )
+      );
+    }, this, transferEncoding);
+  }
+  this.hash = cryptoLib.createHash('md5');
+  this.providedChecksum = typeof Buffer.allocUnsafe === 'function' ?
+    Buffer.allocUnsafe(CHECKSUM_BYTE_LEN).fill(0) :
+    new Buffer(CHECKSUM_BYTE_LEN);
+  this.providedChecksumLen = 0;
+  this.expectedDataLength = contentLength - CHECKSUM_BYTE_LEN;
+  this.receivedLen = 0
+}
+
+/**
+ * Override the flush function in Stream.Transform. It check whether the hashing is correct before
+ * 'end' event emitted and after all data has been transformed and output.
+ */
+IntegrityCheckerStream.prototype._flush = function(callback) {
+  if (
+    this.providedChecksum.length !== CHECKSUM_BYTE_LEN ||
+    this.hash.digest('base64') !== this.providedChecksum.toString('base64')
+  ) {
+    this.emit('error', AWS.util.error(
+      new Error('Response fails integrity check.'),
+      { code: 'ResponseChecksumMismatch' })
+    );
+  }
+  callback();
+}
+
+/**
+ * Override the transfer function in Stream.Transform. It basically slice the last 16 bytes
+ * of the whole response as checksum and the first part as the real data that responsed.
+ */
+IntegrityCheckerStream.prototype._transform = function(chunk, encoding, callback) {
+  if (chunk.length > 0) {
+    if (this.receivedLen >= this.expectedDataLength) {
+      this.providedChecksum.fill(
+        chunk, this.providedChecksumLen, this.providedChecksumLen += chunk.length
+      );
+      chunk = null;
+    } else if (this.receivedLen + chunk.length > this.expectedDataLength) {
+      var newData = chunk.slice(0, this.expectedDataLength - this.receivedLen);
+      var newChecksum = chunk.slice(this.expectedDataLength - this.receivedLen);
+      newChecksum.copy(this.providedChecksum, this.providedChecksumLen);
+      this.hash.update(newData);
+      chunk = newData;
+    } else {
+      this.hash.update(chunk);
+    }
+    this.receivedLen += chunk ? chunk.length : 0;
+  }
+  callback(null, chunk);
```

Comment:
Have you tested to see what happens if the `chunk` passed in contains only the trailing checksum, but not the entire thing? I suspect this would be an edge case but will _transform continue to be called if there's still more data upstream to read, but you've already sent `null` to the downstream stream?

Reply:
Do you mean the case when the first arriving chunk is checksum but the data chunks haven't arrived yet? Is this case even possible?

Reply:
I mean something like the following:
1st time this is called, `chunk` contains just the object's data.
2nd time this is called, `chunk` contains 14 bytes of the trailing checksum, only.
3rd time this is called, `chunk` contains the remaining 2 bytes of the trailing checksum.

I wonder if that last transform call would happen if we already sent `null` in the 2nd call. To be fair, I'm not sure if this is a realistic case. If node buffers a minimum number of bytes (until the end of the stream) before emitting that it's ready, then this is likely overkill to consider. If it can happen, this could end up as a frustrating edge case.